### PR TITLE
ARROW-6211: [Java] Remove dependency on RangeEqualsVisitor from ValueVector interface

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-6211](https://issues.apache.org/jira/browse/ARROW-6211).

This is a follow-up from https://github.com/apache/arrow/pull/4933

public interface VectorVisitor<OUT, IN, EX extends Exception> {..}
n ValueVector : 

public <OUT, IN, EX extends Exception> OUT accept(VectorVisitor<OUT, IN, EX> visitor, IN value) throws EX;

### CODE REVIEW

**Comment by tianchen92** [comment]
cc @pravindra 
I removed Ex from the generic visitor since we have too many places to try-catch exception, and currently RangeEqualsVisitor dosen't need throw exception explicitly.

File: java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java
Comment (pravindra) [inline java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: can we add a API here

equals(ValueVector left) { return visit(left, null); }

that way, the callers don't need to pass null.

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: Seems the caller calls ValueVector#accept, visit() is not called explicitly by caller, so add equals is unnecessary?

Comment (pravindra) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: i was trying to simplify the call in DictionalXX files. I'll get back on this.

Comment (pravindra) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: if we add an API
equals(ValueVector left) { return left.accept(this, null); }

we can change this call 
dictionary.accept(new VectorEqualsVisitor(that.dictionary), null);
to
(new VectorEqualsVisitor(that.dictionary)).equals(dictionary)

Comment (pravindra) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: same for all of the usages in TestValueVector and TestRangeValueVector

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: Sounds good to me, And I think it's blocked by https://github.com/apache/arrow/pull/5100 (if we do like this, it will throw UnsupportedException) opened by @liyafan82 , please take a look if we could get that one merged first.

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: I updated this PR, please help take a look, thanks :)

Comment (pravindra) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: @tianchen92 - why do you need the instanceof checks ? This should work, right ?

public boolean equals(ValueVector left) {
    return left.accept(this, null);
 }

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: I guess not, left.accept(this, null) will call accept(ValueVector left, Void value)
1. if we keep accept(ValueVector) it will throw UnsupportedException
2. if we remove accept(ValueVector) it has compatible error
To solve this, we could either check instanceof here or just in accept(ValueVector). What do you think?

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: @pravindra Hi, could you please take a look at this one when available, I have two following up PRs which are based on this one :)  Thanks

Comment (pravindra) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: I tried replacing the fn with 

public boolean equals(ValueVector left) {
  return left.accept(this, null);
}

all of the tests in memory/vector passed. so, I'm missing something ?

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java]: Sorry I got it wrong, fixed now, thanks!

**Review by pravindra — APPROVED** [review]
@tianchen92 - i'll merge after the tests complete.

**Comment by pravindra** [comment]
thanks for the change, @tianchen92

**Pull request closed without merging**
